### ISSUE TITLE
clear message queue between rounds

### ISSUE THREAD

**Issue by agnescameron** [Kallirroi/bot_test#34]
currently, if you spam the bot by typing too fast, the queued messages get stored and then append to the chat next round.

I tried just setting botQueue to [] each time the bot changes, but i think that the messages are getting stuck in the timeout in Chat.js. Have verified that in this instance the POST request delay isn't the issue (they're returning on average about 145ms), though we might also want to make sure that that isn't a problem for slower internet speeds...

**Comment by agnescameron** [comment]
How to get the effect:
![Screen Shot 2020-03-26 at 14 59 04](https://user-images.githubusercontent.com/16444898/77685815-67ffe580-6f72-11ea-888d-31446d853c5d.png)

Result the following round:
![Screen Shot 2020-03-26 at 14 59 17](https://user-images.githubusercontent.com/16444898/77685814-67ffe580-6f72-11ea-961d-49c0bdc1d3b3.png)

**Comment by cosmicespresso** [comment]
~I have tried replicating this but can't :/~

UPDATE: I see the issue, but its not a problem of clearing the bot queue - it is reset on every round.

**Comment by agnescameron** [comment]
haha perhaps something fixed it... i'll have a go  now

**Comment by agnescameron** [comment]
hmmm. ok I'll do some detective work and get back to you

**Comment by agnescameron** [comment]
<img width="1060" alt="Screen Shot 2020-04-02 at 12 24 00" src="https://user-images.githubusercontent.com/16444898/78273470-dbf33e00-74dc-11ea-980a-a679f919347e.png">
hm -- bot queue is not getting cleared between rounds: here it is still full from the round before. where's it being cleared at the moment?

**Comment by cosmicespresso** [comment]
looking into it - for the moment I'm wondering if all the `if` statements can be handled more elegantly to avoid time lags?

![image](https://user-images.githubusercontent.com/10440556/78273692-31c7e600-74dd-11ea-9639-e3385f093254.png)

**Comment by agnescameron** [comment]
oh yeah for sure, I'll look at that now

**Comment by cosmicespresso** [comment]
![image](https://user-images.githubusercontent.com/10440556/78274207-d8ac8200-74dd-11ea-83ab-ea501eadba13.png)

**Comment by cosmicespresso** [comment]
I **think** the clearing is happening on line 75 of App.js `this.botQueue = this.botQueue.concat(messages);`

**Comment by agnescameron** [comment]
hm, but `concat` doesn't clear anything?

**Comment by cosmicespresso** [comment]
`messages` is an empty array on every new round - but yeah true it shouldn't clear it 
I think its a timing issue (the bot will chew through the backlog at some point and as long as the user is seeing Round [X] screen, they are not typing anything, which gives the bot time to go through the queue. 

I will just try adding a timeout to the typing function and not let the user type too fast?

**Comment by agnescameron** [comment]
is there not another way to clear it? the user timeout typing would have to be pretty long to prevent this from happening

**Comment by cosmicespresso** [comment]
I will look into both, but I think its more important to not put the bot in that position to begin with, rather than clear the queue.

**Comment by agnescameron** [comment]
ya that's true, maybe a combination would be good.

**Comment by cosmicespresso** [comment]
![image](https://user-images.githubusercontent.com/10440556/78275523-b3207800-74df-11ea-9f58-108681cd617c.png)
can you explain whats happening here?

**Comment by agnescameron** [comment]
yes: I think I should move this into text processing though.
it sends this to the text processing middleware, and checks to see if the parser returns a reply (e.g., a bot response that doesn't need to be sent to dialogflow). If there's nothing back from the preprocessor, then the sample is sent to dialogflow

**Comment by agnescameron** [comment]
the context line is a nasty hack (it's actually a bit less nasty in the line i just pushed), at some point we should talk about what your philosophy with the 'state' object is. basically for now it tells the preprocessor whether to activate certain checks or not.

**Comment by cosmicespresso** [comment]
is it possible the callbacks introduce some confusion/time lag? looking at the `await` mostly

**Comment by cosmicespresso** [comment]
also re: StateMap - yes I will finalize that object as soon as we are done with UI updates, and then you can point to specific steps to bind to specific bots.

**Comment by agnescameron** [comment]
where are you noticing the lag? I'll try adding some timers

**Comment by cosmicespresso** [comment]
I think what happens when I type fast is that the UI is displaying my messages, but it doesn't reach the bot as fast as it gets displayed - hence, the bot queue growing. Once I leave some time for the backend to catch up, then that queue is being reduced. That's why I thought that callbacks and `awaits` could be behind this. 

Basically: I'm not sure what regulates how quickly my (from the UI) text reaches the bot, how long it spends there, and how it comes back to the frontend. I thought that the lines I posted above are doing that ?

**Comment by cosmicespresso** [comment]
I think taking a look at what is happening in Network will be useful - not sure what the correct order should be? 
![image](https://user-images.githubusercontent.com/10440556/78277390-6f7b3d80-74e2-11ea-8511-ff492f05dfa4.png)

**Comment by agnescameron** [comment]
ok i think i know what's going on:

when preprocessor runs in non-truth-challenge mode, it takes around 3ms to run: so that’s not the issue

when it parses and returns something without setting context, around 10ms. this is also pretty speedy -- and remember, after that there's no call to the bot (making this the quickest response in terms of total time)

the thing that takes the time is when it sets a context for the bot. this is a callback that takes up to 500ms to return (average about 100-150), so by the time it has returned, then a sample sent to the bot, it can take a long time. 

unfortunately the current way this works means that, a context gets set, then the bot handles the callback. so we couldn’t just set the context then run the response.

BUT — what we could do is move those decisions off the bot entirely. if we’re setting the context in response to user input anyway, we probably know what the intent is: we’re actually using this as a proxy. so, instead of doing that, we can just randomise a response from the bot (even set a context based on response instead!!), which should speed stuff up massively.

the reason you were seeing such a big slowdown, that I hadn’t really noticed when I was debugging the bot queue, is that the only bot context where this currently happens (truth-challenge-bot) was set to the default bot in stateHelpers. In particular, there's a context that gets set when you don't ask the bot a question in the initial interaction. This is now changed.

tl;dr — currently there are some requests where there’s 2 callbacks to the bot; one to set context, and one to get a response. this can be gotten rid of! things will get faster!
